x/auth/src/signing/renderer/tx/tx.rs

```diff
@@ -0,0 +1,451 @@
+use proto_messages::cosmos::{
+    base::v1beta1::SendCoins,
+    tx::v1beta1::{
+        message::Message,
+        public_key::PublicKey,
+        screen::{Indent, Screen},
+        signer_data::SignerData,
+        tip::Tip,
+        tx_data::TxData,
+        tx_metadata::Metadata,
+    },
+};
+use proto_types::{AccAddress, Denom};
+
+use proto_messages::cosmos::ibc::protobuf::Protobuf;
+use tendermint::informal::chain::Id;
+
+use crate::signing::{
+    hasher::hash_get,
+    renderer::value_renderer::{
+        DefaultPrimitiveRenderer, Error, PrimitiveValueRenderer, TryPrimitiveValueRenderer,
+        TryPrimitiveValueRendererWithMetadata, ValueRenderer,
+    },
+};
+
+/// Envelope is an internal data structure used to generate the tx envelope
+/// screens. Used in the same way as the Cosmos SDK Envelope type:
+/// https://github.com/cosmos/cosmos-sdk/blob/main/x/tx/signing/textual/tx.go
+pub struct Envelope<M> {
```

Comment:
Interesting. I removed Envelope when worked on this

Reply:
I think it's useful to have the envelope type to emphasize how transactions are structured for textual signing.